### PR TITLE
Include line info for steps that raise exceptions

### PR DESCRIPTION
**Pull request by philomates**
An attempt to help address https://github.com/nubank/state-flow/issues/74

Given a flow that raises an exception, like:
```clojure
(ns state-flow.scratch
  (:require [state-flow.core :as state-flow]
            [state-flow.cljtest :refer [match?]]
            [state-flow.state :as state]))

(defn inc-a [s]
  (update s :a inc))

(defn change-state-buggy []
  (state/modify inc-a))

(state-flow/run!
  (state-flow/flow "buggy flow"
                   (change-state-buggy)
                   (match? "never run" 1 2)))
```

the result used to be like the following, which doesn't help pinpoint that `(change-state-buggy)` is the offending step:

```
                state-flow.state/reify/fn       state.clj:   63
                 state-flow.scratch/inc-a     scratch.clj:    7
                      clojure.core/update        core.clj: 6196
                         clojure.core/inc        core.clj:  927
                                      ...
java.lang.NullPointerException:

Syntax error (ExceptionInfo) compiling at (state_flow/scratch.clj:12:1).
Flow "buggy flow (scratch.clj:13)" failed with exception
```

With the changes in this PR:

```
                state-flow.state/reify/fn       state.clj:   63
                 state-flow.scratch/inc-a     scratch.clj:    7
                      clojure.core/update        core.clj: 6196
                         clojure.core/inc        core.clj:  927
                                      ...
java.lang.NullPointerException:

Syntax error (ExceptionInfo) compiling at (state_flow/scratch.clj:12:1).
Flow "buggy flow (scratch.clj:13) -> (change-state-buggy) (line 14)" failed with exception
```

**Note:**

we don't have line metadata info for variable references, so

```clojure
(state-flow/run!
  (state-flow/flow "buggy flow"
                   change-state-buggy))
```

results in 
```

Syntax error (ExceptionInfo) compiling at (state_flow/scratch.clj:12:1).
Flow "buggy flow (scratch.clj:13) -> change-state-buggy" failed with exception
```

### CODE REVIEW

File: src/state_flow/core.clj [hunk 1]
```diff
@@ -75,14 +75,46 @@
...
   (let [flow-meta caller-meta
-        flows'    (or flows `[(m/return nil)])]
+        but-last-flows (butlast flows)
+        last-flow      (or (last flows)
+                           `(m/return nil))]
     `(m/do-let
       (push-meta ~description ~flow-meta)
-      [ret# (m/do-let ~@flows')]
+      ~@but-last-flows
+      ~(push-abbr-meta last-flow)
```
Comment (philomates) [inline src/state_flow/core.clj]: adding both the outer flow desc and the actual call sight info can be very useful. For example

```clojure
(state-flow/run!
  (state-flow/flow "buggy flow"




                   (change-state-buggy)))
```

correctly shows
```
Syntax error (ExceptionInfo) compiling at (state_flow/scratch.clj:12:1).
Flow "buggy flow (line 13) -> `(change-stat...)` (line 18)" failed with exception
```

**Comment by sovelten** [comment]
You can also do `(flow "my-buggy-flow" (change-state-buggy))` whenever you need for debugging purposes and you would get correct line numbering then.

It seems like a better trade off than a complicated solution just to provide line number tracking for primitive steps.

**Comment by caioaao** [comment]
This is a more contrived issue. I came across an interesting article about it, though, and might be a good direction to follow: https://pepeiborra.wordpress.com/2009/11/01/monadic-stack-traces-that-make-a-lot-of-sense/

I've only skimmed through the article, but the proposed solution is basically implementing an error monad that carries the stack trace

**Comment by philomates** [comment]
I've cleaned the implementation and it should be ready to review again

File: src/state_flow/core.clj [hunk 2]
```diff
@@ -121,6 +144,28 @@
...
+  `(push-meta ~(description/abbr-sexpr flow)
+              ~(assoc (meta flow)
+                      :call-site-meta true)))
+
+(defn- flow-expr? [expr]
+  (and (coll? expr)
+       (or (= 'flow (first expr))
+           (= `flow (first expr)))))
+
+(defn annote-with-line-meta [flows]
```
Comment (sovelten) [inline src/state_flow/core.clj:157]:
```suggestion
(defn annotate-with-line-meta [flows]
```

File: test/state_flow/core_test.clj [hunk 3]
```diff
@@ -217,23 +209,24 @@
...
-                            level-1-line))))))
+          (is (<= line-number-before-flow-invocation
+                  level-3-line
+                  level-2-line
+                  level-1-line))))))
 
   (testing "after nested flows complete"
     (testing "within nested flows "
-      (is (re-matches #"level 1 \(core_test.clj:\d+\)"
+      (is (re-matches #"level 1 \(core_test.clj:\d+\) -> \(state-flow\/current-description\) \(line \d+\)"
```
Comment (gabrielgiussi) [inline test/state_flow/core_test.clj:219]: If I run this I get
```
=> "level 1 (core_test.clj:2) -> (state-flow/current-description) (line 4)"
```

Where `line 4` is coming from?

**Comment by gabrielgiussi** [comment]
I couldn't find any test when a flow is throwing an exception and the line is included. I guess we are testing `current-description` instead and probably we use `current-description` when an exception is caught but then are not relaying on internal impl details? 🤔 

Shouldn't we have a test covering the case where an exception is raised and we have line information?

**Comment by philomates** [comment]
Thanks for the suggestion @gabrielgiussi, updated with a test

**Pull request merged**